tensorflow/compiler/xla/service/gpu/gpu_executable.cc

```diff
@@ -367,6 +367,18 @@ Status ExecuteThunks(const std::string& module_name,
         async_comms_stream.ok() ? async_comms_stream->get() : nullptr};
     TF_RETURN_IF_ERROR(thunk->ExecuteOnStream(thunk_params));
   }
+
+  // Synchronize async_comms_stream to avoid race condition with temporary
+  // buffer teardown when the allocator allows asynchronous deallocation (e.g.
+  // bfc_allocator).
+  if (async_comms_stream.ok() &&
+      run_options->allocator()->AllowsAsynchronousDeallocation()) {
+    TF_RETURN_WITH_CONTEXT_IF_ERROR(
+        async_comms_stream->get()->BlockHostUntilDone(),
+        absl::StrFormat("Failed to complete all kernels launched on stream %p",
+                        async_comms_stream->get()));
+  }
+
   return MaybeSyncAndProfile(run_options, start_micros,
```

Comment:
Isn't the async comm stream always synchronized with the main stream (as a part of NcclCollectiveDoneThunk)? In which case at the end of the execution, blocking host on the main stream (if `block_host_until_done` is true) should be sufficient?

Reply:
Yes, sync on main_stream here also works. The reason I'm using async_comms_stream is due to its lesser scope given the race condition is between async_comms_stream and buffer tear down.